### PR TITLE
[ONNX][MLIR] Add support for onnx.gather op

### PR DESCRIPTION
**Pull request by Shukla-Gaurav**
This commit adds support for gather op in the onnx pipeline.
https://github.com/nod-ai/SHARK-Turbine/issues/242

### CODE REVIEW

**Comment by Shukla-Gaurav** [comment]
@stellaraccident @rsuderman @vivekkhandelwal1 @renxida Can you please review this PR?

**Review by vivekkhandelwal1 — COMMENTED** [review]
Hi @Shukla-Gaurav, it would be better if you can add a small comment explaining your approach since this is not a straight-forward lowering, and it would be hard for us to understand this by ourselves and then review.

**Comment by Shukla-Gaurav** [comment]
@vivekkhandelwal1 Sure, will add a comment explaining the algorithm in the lowering code itself. Meanwhile you can review it by checking the algorithm here https://github.com/nod-ai/SHARK-Turbine/issues/242

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 1]
```diff
@@ -323,6 +322,131 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+          indexShapeTensor.emplace_back(indexDimVal);
+          prod = rewriter.create<Torch::AtenMulIntOp>(loc, prod, indexDimVal);
+        }
+
+        SmallVector<Value> collapsedIndexSize(dataRank, constOne);
+        collapsedIndexSize[axis] = prod;
+        auto collapsedIndexSizeList =
+            rewriter.create<Torch::PrimListConstructOp>(
+                loc,
+                Torch::ListType::get(Torch::IntType::get(indices.getContext())),
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:387]: You can also do
```
rewriter.getType<Torch::ListType>(rewriter.getType<Torch::IntType>());
```

**Review by rsuderman — CHANGES_REQUESTED** [review]
+1 on having the high level description. I can see several constructions being performed but not interpret how this generates a `onnx.gather` behavior.

**Comment by Shukla-Gaurav** [comment]
@rsuderman @vivekkhandelwal1  
Added high level description of the conversion. Could you please review again? Thanks!

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 2]
```diff
@@ -323,6 +322,154 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+        }
+        ArrayRef<int64_t> indexShape = indexType.getSizes();
+        unsigned indexRank = indexShape.size();
+
+        // 3. Compute total elements in the indices tensor, as we will collapse
+        // the indices tensor to a unary tensor. Also compute index shape and
+        // data shape tensors as they will be used for creating output types.
+        int64_t indexElemCount = 1;
+        for (int64_t dim : indexShape) {
+          if (dim == -1) {
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:360]: We should use `Torch::kUnknownSize` instead

**Review by rsuderman — APPROVED** [review]
Just some minor nits but LGTM otherwise

**Comment by AmosLewis** [comment]
Gather failed again in Shark-TestSuites onnx model opt-125M-awq:
`python ./run.py --torchmlirbuild ../../torch-mlir/build --tolerance 0.001 0.001 --cachedir ./huggingface_cache --ireebuild ../../iree-build  -f onnx -g models --mode onnx --report  --torchtolinalg --tests onnx/models/opt-125M-awq
`
```
opt-125M-awq.default.torch-onnx.mlir:1123:13: error: 'tensor.expand_shape' op expected number of static shape dims to be equal to the output rank (3) but found 0 inputs instead
    %1119 = torch.operator "onnx.Gather"(%0, %1113) : (!torch.vtensor<[50272,768],f32>, !torch.vtensor<[?,?],si64>) -> !torch.vtensor<[?,?,768],f32> 
            ^
opt-125M-awq.default.torch-onnx.mlir:1123:13: note: see current operation: %1049 = "tensor.expand_shape"(%1046) <{reassociation = [[0, 1], [2]], static_output_shape = array<i64>}> : (tensor<?x768xf32>) -> tensor<?x?x768xf32>

```

**Pull request merged**